### DISCUSSION TITLE
I don't like the following designs

### DISCUSSION THREAD

**Discussion by romanholidaypancakes** [carbon-language/carbon-lang#1440] [General]
1. Why do we need to type more characters `fn` and `->`, obviously many times in one fell swoop
```
 i32 main()
```
best than
```
fn main() -> i32
```

2. Are `var` and `:` also redundant?

```
i64 i = -1;
```
best than
```
var i:i64 = -1;
```

**Comment by Turbine1991** [comment]
As somebody who finds it uncomfortable in C++, this syntax I'm almost willing to accept, due to the inconsistencies.

1) I like that is faster for the compiler, as it doesn't have to determine something's a type before realising it's a function.
1) You can more clearly align variable/function names without the type sending it flying across the screen.
2) `var i:i64 = -1;`
Yeah, this kind of inconsistency is not very nice. -> or : but not both. Delphi used the :

**Reply by jwtowner** [reply]
Number 2 is important. There are a many C++ code bases out there where the developers continually waste time tab-indenting function declarations so that the function names line up. The function name is more important than the return type.

**Reply by kevin-j-channon** [reply]
As mentioned, the trailing return type syntax is simpler to parse. We also have trailing return types in C++; you can write:
```
auto my_func() -> int32_t
```

**Reply by Twins-Divyanshu-Sharma** [reply]
if you want to write a compiler friendly language, write the code in assembly

**Reply by jonmeow** [reply]
@Twins-Divyanshu-Sharma Be kind and courteous on our forums. Please review our code of conduct. https://github.com/carbon-language/carbon-lang/blob/trunk/CODE_OF_CONDUCT.md

**Comment by pache-Ak** [comment]
I only agree the trailing return type consider generics 
` auto f(T t) -> T::value_type {} `
return type depend on parameter type.  classic syntax is trouble.
I suggestion syntax like this :
`func` *name*`(` *args* `)` *return type* `{  }`
but not be accepted.

I can accept keyword `var`  complier need know a name is  type, but I don't like `:` , `var` *type* *name* is too much, I don't like more things.

**Comment by Twins-Divyanshu-Sharma** [comment]
why use 
`var name : i32`
why not
`name : i32`
or
`var name`

it is a strongly typed language with a ILLUSION of weakly typed.

**Comment by jonmeow** [comment]
@inaryart I'm changing the title of this. The title was using a harsh way of expressing your opinion. Please bear in mind Carbon is not for everyone, but we do expect everyone to be kind and courteous. Review our code of conduct: https://github.com/carbon-language/carbon-lang/blob/trunk/CODE_OF_CONDUCT.md

**Reply by jordirosa** [reply]
If he was harsh i don't agree with the manners, but i strongly agree with the constructive side of the message and i think it's very important, you say carbon is not for everyone, but if as said in readme carbon intend to be a C++ successor you're probably targeting at C++ developers, so why to change all the syntax developers already know even when this is not needed because a missing feature of C++? The new syntax does not improve the old one, it has the same elements but written in a different way, so is not even simpler as said in the readme and will make this new language less likely to be adopted for your target users.

I think the new syntax is a matter of taste someones will like it and someones will not, but if the goal of the project is to be adopted by C++ developers i think C-like syntax would have been a much better approach.

**Reply by Turbine1991** [reply]
Totally agree. If you're trying to appeal to c++ devs, at least keep the good. Otherwise you're shooting your potential userbase. I've already decided this project is not for me as they've [made too many wrong choices]

**Reply by Zaya6** [reply]
> Totally agree. If you're trying to appeal to c++ devs, at least keep the good. Otherwise you're shooting your potential userbase. I've already decided this project is not for me [...]

I've more or less come to the same conclusion. Judging by the replies I've seen and the general outlook of commentators on this project about anything c++ related. This doesn't appear like a project for c++ devs. This is a project for outsiders who dislike the language and want to see it become more like typescript or go. I get this feeling because I used to be one of them. Then I actually learned the language and realized non of it is as bad as the internet portrays. The c-syntax is not the thing c/c++ programmers need fixed besides maybe better consistency.

**Reply by geoffromer** [reply]
I've edited @Turbine1991's comment to replace phrasing that's not consistent with our [code of conduct](https://github.com/carbon-language/carbon-lang/blob/trunk/CODE_OF_CONDUCT.md) (especially "be kind" and "be friendly and patient") and edited @Zaya6's comment to remove a quotation of it. The edits are marked with square brackets; feel free to make further edits if you feel my paraphrase is inaccurate, but your edits (like all postings in Carbon project forums) must comply with the code of conduct.

**Reply by jordirosa** [reply]
@Zaya6 I guess you’re right but then they’re not a successor but another alternative, and they are catching the attention of wrong kind of devs with the claim.

**Comment by necutli** [comment]
current syntax to declare
fn main() -> i32 { }
fn main()  { }
var i:i64 = -1;


what about next
fn i32 main()  {  }
fn main()  {  }
var i64 i  = -1;

**Reply by Zaya6** [reply]
I'm on board with this. Unlike others I really can't stand the extra keystrokes everyone is adding. This is much more c-like to me and my wrists will be thankful

**Reply by vbregier** [reply]
Please understand that code is read orders of magnitude more than it is written.
What matters is not the time you need to spend to write code, but the effort it is to read it !

**Comment by vbregier** [comment]
I disagree with this topic.
I do like the design, new syntax is more readable, therefore better, than c++ syntax.

(though I would prefer func or function instead of fn, and int32 instead of i32, but there are other topics for these keywords)

**Comment by Wajinn** [comment]
Why not use the walrus operator (`:=`), as do several languages?

`var i:i64 = -1`;

If the value is omitted then you have a declaration without an initialization:

`var i := i64`;

If the type is omitted then can be inferred based on value:

 `var i := -1`;

Just seems like `var i:i64 = -1` is leading to it.  The above suggestions, using the walrus operator, can also still work with the existing syntax and being a shorter way.

**Comment by DevNoteHQ** [comment]
Well the idea is probably to have type inference at some point. All modern languages use it in one way or another, so carbon should obviously have it to. And with this syntax, you can just skip the type definition instead of having to write auto.
So something like `var i = -1;` instead of `auto i = -1;` like in C++ or even `var i: auto = -1`.
This combines the benefit of simplicity in the definition (as you can simply omit the type) while beeing very explicit on whether the variable is readonly or not. And it also makes it very easy for the compiler to identify variables since you have a dedicated keyword for it (technically two with `let`).

**Reply by jonmeow** [reply]
The planned syntax is `var i: auto`; the trade-offs of eliding in particular are discussed a bit here:
https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0851.md#elide-the-type-instead-of-using-auto

**Reply by DevNoteHQ** [reply]
I've seen that, but I kind of don't understand how those disadvantages are considered to be relevant?
I would understand it if carbon would stick closer to C++ and do something like `auto i = -1`.
If you omit the type here `i = -1`, you can't be sure if it's a new type declaration or an assignment and type inference becomes the lack of a type rather than the precense of a infered type.

The main benefit of requiring `var` in variable definition imo is to be able to omit the type without loosing the information that you are trying to define a variable here. Java even introduced the `var` keyword instead of `auto`, so that you can also write `var i = -1` which has a similar meaning to C++'s `auto i = -1`.
Sticking to auto just combines the disadvantages of both C++ like syntaxes and more script like syntaxes. You explicitly have to type `var` and `auto` to have a type infered variable, where infered types have imo the main benefit of beeing faster to write. With shorter types, there is simply no benefit in using auto instead of the type `var i: auto = new Cat()` vs `var i: Cat = new Cat()`. This becomes a lot shorter with omitted types `var i = new Cat()`.

As the proposal already mentions, carbon is the first language to use both a variable declaration keyword while also requiring a type inference keyword. All other languages stick to ommiting the type. Simply because that's the easiest and fastest syntax. Why reinvent the weel here?

So I really hope you reconsider.

**Reply by geoffromer** [reply]
> I would understand it if carbon would stick closer to C++ and do something like `auto i = -1`.
> If you omit the type here `i = -1`, you can't be sure if it's a new type declaration or an assignment and type inference becomes the lack of a type rather than the precense of a infered type.

Similar reasoning actually applies in Carbon, because the syntax between `var` and `=` isn't necessarily a single variable declaration, it's a pattern, and patterns can contain expressions. For example, `(x: i32, y: i32)` is a pattern that matches any tuple of type `(i32, i32)` and binds the names `x` and `y` to its elements, whereas `(x: i32, y)` is a pattern that matches any 2-tuple whose second element is equal to `y`, and binds `x` to its first element. So we need some syntax that distinguishes a new variable declaration from a use of an existing variable.

**Reply by DevNoteHQ** [reply]
Well then just enforce `auto` just for patterns? Since there is no `var` involved in declaring the fields, there is really no other option then to require either a dedicated type or `auto`. But that is simply not the case for variable declarations. And variable declarations are probably the most used feature of any programming laguage, so saving a keyword here should really pay off.



But having now found the design documentation (sorry, it's only linked at the bottom of the main readme, should be at the top imo) I have to say that I find the `match`-syntax a bit scatchy anyway. Lets say we have `fn test() -> (i32, i32) { return (4, 5) }` and then call that in a `match` like:
```
match(test()) {
     case (4, y: i32) => ...
     case (x: i32, 5) => ...
}
```
Will both cases then be executed? That seems to be quite unsafe. If both change the same variable, then the last executed statement will overwrite all the other ones.
Now I would expect a `switch`-like statement like `match` to only execute one option at maximum. So only the first one to match will be executed? That's the way Kotlin handels it with two "matches" in a `when` expression. But it's much harder to do this in Kotlin (at least given one "input" variable) since the whole object has to match the given expression, not just parts of it.

**Reply by mpingleton** [reply]
I agree that writing 'auto' in Carbon would be overly redundant.  The whole reason the 'auto' keyword was invented for C++ in the first place was to enable type-inference in an older style syntax that used to not support it.  I'd argue that the 'auto' keyword as a whole would be pointless for any modern programming language, including Carbon.

**Reply by DevNoteHQ** [reply]
Agreed. Though if Carbon sticks to refutable patterns then I would agree with @geoffromer that the auto keyword is required to distinguish values from type-infered variables like:
```
var y = 5 \\ No ": auto" required here if type inference is possible
match (mySuperFunction()) {
     case (3, y \* y is a value here \*) -> ...
     case (5, y: auto \* y is a variable here, should probably use a different variable name to evade conflicts with the first y \*) -> ...
}
```

**Reply by geoffromer** [reply]
> Well then just enforce `auto` just for patterns? Since there is no `var` involved in declaring the fields, there is really no other option then to require either a dedicated type or `auto`. But that is simply not the case for variable declarations. And variable declarations are probably the most used feature of any programming laguage, so saving a keyword here should really pay off.

`auto` is only required (or even allowed) in patterns, but the syntax that follows `var` is a pattern. For example:

```
var (x: i32, y: i32) = foo();
```

Conceivably we could allow `: auto` to be omitted in contexts where the pattern is required to be irrefutable, but that would make the language less consistent, so it's not clear that would be a net improvement

> But having now found the design documentation (sorry, it's only linked at the bottom of the main readme, should be at the top imo) I have to say that I find the `match`-syntax a bit scatchy anyway. Lets say we have `fn test() -> (i32, i32) { return (4, 5) }` and then call that in a `match` like:
> 
> ```
> match(test()) {
>      case (4, y: i32) => ...
>      case (x: i32, 5) => ...
> }
> ```
> 
> Will both cases then be executed? That seems to be quite unsafe. If both change the same variable, then the last executed statement will overwrite all the other ones. Now I would expect a `switch`-like statement like `match` to only execute one option at maximum. So only the first one to match will be executed? That's the way Kotlin handels it with two "matches" in a `when` expression. But it's much harder to do this in Kotlin (at least given one "input" variable) since the whole object has to match the given expression, not just parts of it.

A `match` statement will execute the first case that matches.

**Reply by OlaFosheimGrostad** [reply]
I think the designers should focus on the core language first, sans syntax sugar. So it is ok if things are a bit longwinded now, as long as they make sure that it can be fixed later with sugar.

**Reply by DevNoteHQ** [reply]
> Conceivably we could allow : auto to be omitted in contexts where the pattern is required to be irrefutable, but that would make the language less consistent, so it's not clear that would be a net improvement

Well refutable patterns are quite a bit different in their meaning/usage from irrefutable patterns, so I don't see a real problem in having different syntax for them. I would even suggest to have separate types of parentheses for them to make them more distinguishable.
```
var (x: i32, y: i32) = foo();

match(foo()) {
     case [4, y: i32] => ...
     case [x: i32, 5] => ...
}
```
Though as I said I would prefer a more Kotlin-`when`-like syntax for `match`.

> I think the designers should focus on the core language first, sans syntax sugar. So it is ok if things are a bit longwinded now, as long as they make sure that it can be fixed later with sugar.

Yeah that's probably true. Omitting the `auto` keyword shouldn't really be a challenge even when it's just implemented shortly before 1.0 release. But it would be a good idea then to create a pinned "Controversial syntax" mega thread. Otherwise there'll just be the same discussions over and over again^^

**Comment by flysand7** [comment]
The reason why Carbon can't afford the

```
i32 x = 3;
```

syntax is actually quite simple. Doing so violates the following [design principle](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/goals.md#language-tools-and-ecosystem):

> No semantic or contextual information used when parsing

The C/C++ declaration syntax requires you to have the type information at the stage of parsing. Consider a simple parser going through the tokens one by one, trying to figure out what whether it parses an expression or a declaration:

```
   i32 x = 3
   ^^^
   Since no contextual information can be used,
   the parser can't distinguish types from variable
   names therefore it can't deside whether it's parsing
   a declaration or an expression involving variable
   named "i32".
```

You might consider what happens if the parser is allowed to assume that identifier following another identifier is a sign on a declaration:

```
   i32 x = 3
   ^^^^^
   Sounds like a declaration, doesn't seem to have issues
```

However, what if instead we chose a more complex example:

```
   i32 * x = 3;
   ^^^^^
   It looks like pointer to i32, but actually we don't know
   that i32 is a type (because parser has no contextual
   information), so actually it might be a multiplication
   expression of i32 and x.
```

This is a tiny hint of unbounded lookahead. It gets worse once we start considering declarations of functions and pointers to functions. Note that in C and C++ parsers get away with using contextual information to reduce lookahead, which is also the reason why in those languages declaration *must* precede usage.
